architecture/2024-11-authz-feature-spec.md

```diff
@@ -0,0 +1,390 @@
+  --assignee developer@my-company.net \
+  --role developer \
+  --scope /planes/radius/MyCompany/resourceGroup/default-application
+rad role assignment create \
+  --assignee developer@my-company.net \
+  --role deployer \
+  --scope /planes/radius/MyCompany/resourceGroup/default-environment
+```
+
+**Result**
+
+1. `developer@my-company.net` is added to the out-of-the-box role `developer` for the `default-application` resource group 
+2. `developer@my-company.net` is added to the out-of-the-box role `deplorer` for the `default-environment` resource group
+
+**Exceptions**
+
+The operation fails and informs the user interactively if:
+
+* The current user is not a member of the `radius-admin` or `group-admin` roles or the `cluster-admin` RoleBinding in the Kubernetes cluster
+
+### Scenario 3 – Multi-tenant enterprise
+
+In a complex, multi-tenant enterprise environment, Radius will be used by multiple development teams to build and maintain multiple different applications. Infrastructure management duties will be split across multiple infrastructure teams. For example, cloud environments may be maintained by a cloud infrastructure team while on-premises is maintained be a data center team. Resource types and recipes may be maintained by different teams as well. The DBA team may maintain database resource types and recipes while an enterprise architecture team maintains other resource types. 
+
+Large enterprises need a flexible authorization model which enables them to tightly control a granular set of actions permitted on granular set of resources. To describe this scenario in more depth, a series of user stories will be described based on the diagram below.
+
+![](2024-11-authz-feature-spec/authz.png)
+
+#### User story 3 – Create resource group for environments
+
+As a platform engineer at a large enterprise, I need to create a resource group for a cloud environment and grant access to manage environments to my cloud infrastructure team. I have deleted the `default-environment` resource group.
+
+**User Experience**
+
+```bash
+# Create resource group for non-production environments
+rad group create non-prod-env
+# Assign Environment Administrator role to the cloud engineering user group
+rad role assignment create \
+  --assignee cloud-engineering@my-company.net \
+  --role env-admin \
+  --scope /planes/radius/MyCompany/resourceGroup/non-prod-env
+```
+
+**Result**
+
+1. An empty resource group is created called `non-prod-env`
+2. The `env-admin` role is granted to the cloud engineering group to manage environments within the `non-prod-env` resource group
+
+**Exceptions**
+
+The operation fails and informs the user interactively if:
+
+* The current user is not a member of the `radius-admin` or `group-admin` roles or the `cluster-admin` RoleBinding in the Kubernetes cluster
+* The resource group already exists
+* The role assignment already exists
+
+#### User story 4 – Create resource group for an application
```

Comment:
Can you give more detail on how we would grant or revoke the ability to use a resource type namespace?

Reply:
A user is granted the ability to perform operations (actions in this functional spec) on a set of resource types. For example, to create a container a developer would need Applications.Core/containers/* added to their role definition.  Then they would need that role definition assigned to them within a specific scope (i.e., resource group). 

If an admin wanted to restrict usage of Applications.Core/containers, they could remove that from their role definition and replace it with something like MyCompany.App/* which would be the entire resource catalog for MyCompany.App.